driver-core/src/main/com/mongodb/internal/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -82,7 +82,6 @@ public void next(final SingleResultCallback<List<T>> callback) {
             public void apply(final AsyncAggregateResponseBatchCursor<RawBsonDocument> cursor,
                               final SingleResultCallback<List<RawBsonDocument>> callback) {
                 cursor.next(callback);
-                cachePostBatchResumeToken(cursor);
```

Comment:
This is the main culprit - an obvious race, once you know what line to look at.

Reply:
Oh my.